### PR TITLE
Am 859 openid connect implementation

### PR DESCRIPTION
**Pull request by nitishsingla1234**
### JIRA link (if applicable) ###
AM-859

### Change description ###
openid connect implementation

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -271,6 +271,15 @@ dependencies {
...
   compile group: 'org.postgresql', name: 'postgresql', version: '42.2.14'
+  implementation 'org.springframework.boot:spring-boot-starter-oauth2-client'
+  compile "org.springframework.security:spring-security-web"
+  compile "org.springframework.security:spring-security-config"
+  compile "org.springframework.boot:spring-boot-starter-oauth2-client"
+  compile "org.springframework.boot:spring-boot-starter-oauth2-resource-server"
+  implementation 'uk.gov.hmcts.reform:idam-client:1.5.4'
+  compile group: 'com.nimbusds', name: 'nimbus-jose-jwt', version: '7.9'
+  compile group: 'io.jsonwebtoken', name: 'jjwt', version: '0.9.1'
+
```
Comment (aashishaggarwal123) [inline build.gradle:282]: oauth2-client is added 2 times. Can we have all the dependencies for compile only?

File: src/main/resources/application.yaml [hunk 2]
```diff
@@ -71,6 +71,22 @@ spring:
...
+        provider:
+          oidc:
+            issuer-uri: ${OPEN_ID_API_BASE_URI:https://idam-web-public.aat.platform.hmcts.net/o}
+        registration:
+          oidc:
+            client-id: microservicekey-am-role-assignment-service
+            client-secret: ${AM_ROLE_ASSIGNMENT_SERVICE_SECRET:AAAAAAAAAAAAAAAA}
+            scope:
+              - openid
+              - roles
```
Comment (KOTLAPRA23) [inline src/main/resources/application.yaml:86]: i dont think you need scope here?

File: src/main/resources/application.yaml [hunk 3]
```diff
@@ -71,6 +71,22 @@ spring:
...
+              - profile
+
+oidc.issuer: ${OIDC_ISSUER_URL:https://forgerock-am.service.core-compute-idam-aat.internal:8443/openam/oauth2/hmcts}
 
 ### dependency configuration
 auth.provider.service.client.baseUrl: ${IDAM_S2S_URL:http://localhost:4502}
```
Comment (KOTLAPRA23) [inline src/main/resources/application.yaml:92]: this entry also not required

File: src/main/resources/application.yaml [hunk 4]
```diff
@@ -80,8 +96,11 @@ idam:
     totp_secret: ${AM_ROLE_ASSIGNMENT_SERVICE_SECRET:AAAAAAAAAAAAAAAA}
     microservice: am_role_assignment_service
     url: ${IDAM_S2S_URL:http://localhost:4502}
+  s2s-authorised:
+    services: ${ROLE_ASSIGNMENT_S2S_AUTHORISED_SERVICES:ccd_gw,am_role_assignment_service}
+  api.url: ${IDAM_USER_URL:http://localhost:5000}
 
-auth.idam.client.baseUrl: ${IDAM_USER_URL:http://localhost:5000}
+auth.idam.client.baseUrl: ${IDAM_USER_URL:http://localhost:5000/o}
```
Comment (KOTLAPRA23) [inline src/main/resources/application.yaml:103]: this entry aswell not required

File: src/main/java/uk/gov/hmcts/reform/roleassignment/config/SecurityConfiguration.java [hunk 5]
```diff
@@ -1,68 +1,108 @@
...
-            .addFilter(authCheckerFilter)
+            .addFilterBefore(serviceAuthFilter, BearerTokenAuthenticationFilter.class)
             .sessionManagement().sessionCreationPolicy(STATELESS).and()
             .csrf().disable()
             .formLogin().disable()
             .logout().disable()
             .authorizeRequests()
+            .antMatchers("/error").permitAll()
+            //.antMatchers(HttpMethod.POST, "/refdata/external/v1/organisations").permitAll()
+            //.antMatchers(HttpMethod.POST, "/refdata/internal/v1/organisations").permitAll()
```
Comment (aashishaggarwal123) [inline src/main/java/uk/gov/hmcts/reform/roleassignment/config/SecurityConfiguration.java:82]: we can remove these commented references

File: src/main/java/uk/gov/hmcts/reform/roleassignment/config/SecurityConfiguration.java [hunk 6]
```diff
@@ -1,68 +1,108 @@
...
             .anyRequest()
-            .authenticated();
+            .authenticated()
+            .and()
+            .oauth2ResourceServer()
+            .jwt()
+            .jwtAuthenticationConverter(jwtAuthenticationConverter)
+            .and()
+            .and()
```
Comment (aashishaggarwal123) [inline src/main/java/uk/gov/hmcts/reform/roleassignment/config/SecurityConfiguration.java:90]: I think, one and can be removed.

**Pull request merged**